### PR TITLE
METRON-1232: Alert status changes are not reflected in list view

### PR DESCRIPTION
**Pull request by iraghumitra**
## Contributor Comments
The Jira has all the description of the issue. 
In addition, the e2e test for removing search chip was consistently failing provided a small fix for it.

## Steps to reproduce the problem

- Login to alerts UI
- Click on the pause button for alerts polling
- Select an alert, this will open the details pane for the selected alert
- Change the Alerts status by selecting the button on the top (Ex: ESCALATE, NEW, OPEN etc)
- Close the deatils pane by clicking on the close icon
- Thought the status of alert is persisted, the status field in list view is not updated (This is fixed in the PR)

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-interface/metron-alerts/src/app/shared/directives/alert-search.directive.ts [hunk 1]
```diff
@@ -130,7 +130,7 @@ export class AlertSearchDirective implements AfterViewInit, OnChanges {
 
   private handleMouseEvent (callback: Function) {
     clearTimeout(this.mouseEventTimer);
-    this.mouseEventTimer = setTimeout(() => { callback(); }, 500);
+    this.mouseEventTimer = setTimeout(() => { callback(); }, 100);
```
Comment (nickwallen) [inline metron-interface/metron-alerts/src/app/shared/directives/alert-search.directive.ts]: Was this change intentional?

File: metron-interface/metron-alerts/e2e/alerts-list/alerts-list.po.ts [hunk 2]
```diff
@@ -231,10 +231,12 @@ export class MetronAlertsPage {
   }
 
   clickRemoveSearchChip() {
-    let aceLine = element.all(by.css('.ace_keyword')).get(0);
-    browser.actions().mouseMove(aceLine).perform().then(() => {
-      this.waitForElementPresence(element(by.css('.ace_value i'))).then(() => {
-        element.all(by.css('.ace_value i')).get(0).click();
+    element(by.css('app-alerts-list .ace_text-input')).sendKeys('').then(() => {
```
Comment (mmiklavc) [inline metron-interface/metron-alerts/e2e/alerts-list/alerts-list.po.ts]: I'd like to see a couple things with this. Can you add some code comments along with perhaps splitting this up into more descriptive functions? There's a lot of chaining/lambda/closures going on here.

File: metron-interface/metron-alerts/src/app/shared/directives/alert-search.directive.ts (hunk 1 above)
Comment (iraghumitra) [reply metron-interface/metron-alerts/src/app/shared/directives/alert-search.directive.ts]: yeah i feel this is making some of the flaky tests stable, especially the ones where we hover on search text and delete them

Comment (nickwallen) [reply metron-interface/metron-alerts/src/app/shared/directives/alert-search.directive.ts]: hey @iraghumitra - this slipped my mind earlier.  can you point me to the tests for this functionality?  I don't want this merged unless we have good tests.

Comment (iraghumitra) [reply metron-interface/metron-alerts/src/app/shared/directives/alert-search.directive.ts]: @nickwallen The status check was getting checked implicitly when we assert the table data. NowI added a explicit assert as well

File: metron-interface/metron-alerts/src/app/service/update.service.ts [hunk 3]
```diff
@@ -32,13 +33,31 @@ export class UpdateService {
...
+  alertChangedSource = new Subject<PatchRequest>();
+  alertChanged$ = this.alertChangedSource.asObservable();
 
   constructor(private http: Http) { }
 
   public patch(patchRequest: PatchRequest): Observable<{}> {
     let url = '/api/v1/update/patch';
-    return this.http.patch(url, patchRequest, new RequestOptions({headers: new Headers(this.defaultHeaders)}))
-    .catch(HttpUtil.handleError);
+    let observable = Observable.create((observer => {
```
[the remark this answers is not in the document]
Comment (iraghumitra) [reply metron-interface/metron-alerts/src/app/service/update.service.ts]: @merrimanr This is not an anti-pattern as such, we can trigger the event from the component that is trying to patch, but it has to be done at all places wherever we are patching. In this way there is no such need and list view will definitely get intimated of all changes. 
If you are not comfortable we can change it :)

**Comment by iraghumitra** [comment]
@nickwallen @mmiklavc @merrimanr replies are inline in the review

File: metron-interface/metron-alerts/src/app/service/update.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (merrimanr) [reply metron-interface/metron-alerts/src/app/service/update.service.ts]: I'm not suggesting we move the alertChangedSource call to all the subscribing components.  Instead you can move that call to a map function similar to:
```
this.http.patch(url, patchRequest, new RequestOptions({headers: new Headers(this.defaultHeaders)}))
  .catch(HttpUtil.handleError)
  .map(res => {
    this.alertChangedSource.next(patchRequest);
    return res;
  }
 .subscribe(result => reqCompleted(result));
```

[the remark this answers is not in the document]
Comment (iraghumitra) [reply metron-interface/metron-alerts/src/app/service/update.service.ts]: @merrimanr removed the extra observable made change similar to your suggestion

**Comment by nickwallen** [comment]
+1 (again, even!) Pending Travis and resolving feedback from @mmiklavc and @merrimanr

I was able to run the e2e tests successfully and confirmed they do sufficiently test the change.

**Comment by james-sirota** [comment]
+ 1 from me as well pending the completion of the Travis build

**Comment by mmiklavc** [comment]
+1

**Comment by merrimanr** [comment]
+1
